### DISCUSSION TITLE
How do I profile layer-by-layer Jax performance?

### DISCUSSION THREAD

**Discussion by abhinavgoel95** [jax-ml/jax#11180] [Q&A]
I am looking to break down Jax performance, layer-by-layer. Is there a way to place a breakpoint during forward propagation?

**Comment by YouJiacheng** [comment]
Unfortunately, in general JAX might do cross layer fusion, exact performance is hard to measure. However, you can use `host_callback` to measure it roughly: check https://github.com/google/jax/discussions/10908#discussioncomment-2862933
i.e. you can print the time difference between input ready and output ready of a layer, but you cannot ensure there isn't any other operation irrelevant to the layer executed on GPU during this period.

**Reply by abhinavgoel95** [reply]
Thank you for your reply.

**Comment by hawkinsp** [comment]
The main tool we tend to use is the tracing/profiling infrastructure described here: https://jax.readthedocs.io/en/latest/profiling.html

As YouJiacheng says, we don't compile layers in isolation, so it's not always easy to point to a single thing on a trace which exactly corresponds to a layer.

One thing that will certainly work would be to `jit`-compile that layer in isolation and write a microbenchmark that exercises it. But I'd certainly start with the tracing tools.

**Reply by abhinavgoel95** [reply]
Thank you for your reply.

**Reply by sharadmv** [reply]
Also consider using `jax.named_scope` or `jax.named_call` to annotate your layers so they appear in the trace viewer.
